Implement possibility to have sub-devices

derjoerg opened the pull request:
Closes #157 

This implements the possibilities to have sub-devices for each "multi-device" in F@H, which can be placed independently on the floorplan.

This can increase the device-count drastically. In my case it increased from 86 to 181 devices.

The default is false (also for existing installation), so the change needs to be specific wanted by the user.

Additionally, what I saw during my tests: If you switch on this feature, the existing multi-device will preserve the original set area (this needs to be manually removed), only with a fresh installation the physical (multi) device will not have an area assigned.

I also encountered (once again) a small bug from ABB:
E.g. a virtual switch and a virtual energy meter are correctly displayed in HA as one device, but a virtual binary sensor does not has a floor and a room assigned on device level, only on channel level. This means that you will have for a virtual binary sensor a device within a device.

kingsleyadam commented on custom_components/abbfreeathome_ci/binary_sensor.py:
```diff
@@ -148,6 +156,18 @@ def __init__(
             **entity_description_kwargs,
         )
 
+        if self._create_subdevices and self._channel.device.floor is None:
+            device_registry = dr.async_get(hass)
+            device_registry.async_get_or_create(
+                config_entry_id=config_entry_id,
+                identifiers={(DOMAIN, self._channel.device_serial)},
+                name=self._channel.device_name,
+                manufacturer="ABB Busch-Jaeger",
+                serial_number=self._channel.device_serial,
+                suggested_area=None,
+                via_device=(DOMAIN, self._sysap_serial_number),
+            )
```
I think there's maybe a better way to handle this. This is essentially repeat logic on each class, I was thinking we would add a new property to the `Device` class that would determine whether it should be treated as a "multi-device" type.

Then, within the `__init__.py` we can loop through all multi-devices and add them all at once. This avoids having to add this logic on each class. We could take it a step further and add all Devices in `__init__.py` class and just use the `device_info` property within the entity to match-up the entity to the device we've already added (https://developers.home-assistant.io/docs/device_registry_index/#defining-devices). Or in the case of these multi-device entities they'll define their own device within the `device_info`.

Adding a new property to `Device` also gives us flexibility on how a Device is detirmined to be a multi-device. I'm not a huge fan of using the floor info, but that seems to be the only way, but maybe in a future revision of the SysAP firmware, AAB makes them identifiable. Then we only have to change the python package vs having to adjust this logic all over the integration.

Also, having a new property makes it easy for triaging these multi-devices. We can list all the devices from the api and whether they would be treated at multi-device. For example, if I enable this feature I'd like to know which devices will be treated differently within the integration.

derjoerg replied:
> > Then, within the `__init__.py` we can loop through all multi-devices and add them all at once. This avoids having to add this logic on each class. We could take it a step further and add all Devices in `__init__.py` class and just use the `device_info` property within the entity to match-up the entity to the device we've already added (https://developers.home-assistant.io/docs/device_registry_index/#defining-devices). Or in the case of these multi-device entities they'll define their own device within the `device_info`.
> 
> I have to admit, that I did it exactly this way at my first try. But to ease the development I commented out all platforms beside `binary_sensor`. And then I encountered that a lot of "empty" devices were created. So I moved the logic from the integration- to the platform-level. I have totally overseen, that the use-case "empty devices" isn't possible when all platforms are enabled. I will incorporate that.
> 

@kingsleyadam 
This doesn't work. The `load()´-function loads all devices of the defined interfaces. At this point in time we don't know if we have the functionIDs defined.
E.g. I have a "SysAP communication interface" device without any channels. Nevertheless it is in the "devices" list. So when I circle through all devices I also add this device but it will never have any sub-devices and/or entries.

We only know on a per-channel base, if the associcated device is needed or not.

kingsleyadam replied:
> If I understand you correctly, you imagine something like this: As soon as a user activates the "multi-device" option, they receive a message similar to: "The following devices will be switched to 'multi-device':"? If so, I have no plan how to implement that

I'm more thinking along the lines of, here's a script you can run to see how many "multi-devices" you have, make sure you're ok with this.

Here's an example output from my home (I asked Copilot to throw something together). And it's a good representation on why I don't really like using the floor information. It's not a great indicator as there are other scenarios where a device can have a missing floor. But as mentioned before, we don't have many other options if we want this.

In my setup I'll have 1 channel that belongs to a "multi-device" which is the `Panic Device`. I think the `Panic Device` is a bit of a snowflake, special type of device. But in my setup it'd get 2 devices created for the `Panic Device` switch sensor.

So we may want other checks, possible like `number of channels > 1`?

```
                                Multi-Devices Overview                                
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
┃ Device                         ┃ Serial               ┃ Interface       ┃ Channels ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
│ SMART-BRIDGE-0d66511648657080… │ 6000F91624D1         │ VD              │        0 │
│ SMART-BRIDGE-0d66511648657080… │ 600044A55D11         │ VD              │        0 │
│ Everything Off Action          │ FFFF4A010001         │                 │        0 │
│ TV area                        │ BEED509C01C8         │                 │        0 │
│ Panic Device                   │ ABB29E4B4151         │ TP              │        1 │
│ SysAP                          │ ABB700000000         │                 │        0 │
│ SysAP communication interface  │ ABB7F59BF8CF         │                 │        0 │
│ Living room                    │ BEED509C0151         │                 │        0 │
│ Bedroom                        │ BEED509C0153         │                 │        0 │
│ SysAP communication interface  │ ABB7F59BF8D0         │                 │        0 │
│ SMART-BRIDGE-0d66511648657080… │ 6000A17E52E9         │ VD              │        0 │
│ Kitchen                        │ A00073E78ED0         │ sonos           │        0 │
│ 013070529299002421             │ 6000E66F3B4E         │ VD              │        0 │
│ Porch Light Evening            │ FFFF4A000004         │                 │        0 │
│ Kitchen Island                 │ BEED509C0019         │ hue             │        0 │
│ Time program actuator          │ FFFF4A000002         │                 │        0 │
│ Media Player                   │ A0006FB45DB9         │ sonos           │        0 │
└────────────────────────────────┴──────────────────────┴─────────────────┴──────────┘

Panic Device Channels:
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┓
┃ Name                      ┃ Type                 ┃ ID         ┃ Location             ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━┩
│ Panic Device              │ SwitchSensor         │ ch0001     │ Unknown              │
└───────────────────────────┴──────────────────────┴────────────┴──────────────────────┘
```

kingsleyadam replied:
Also, here's the script that produced that output if you want you check your own. You need to install the `rich` PyPi package (`uv pip install rich` if you're in the dev container). I added a new property to device to see what the affect on my setup would be.

```python
    @property
    def is_multi_device(self) -> bool:
        """Return True is this device should be considered a multi-device."""
        return self.room is None
```

```python
import asyncio
from rich.console import Console
from rich.table import Table
from rich.tree import Tree
from rich.panel import Panel
from rich.text import Text

from abbfreeathome import FreeAtHome, FreeAtHomeApi


def display_multi_devices_rich(devices: dict):
    """Display multi-devices using rich formatting."""
    console = Console()

    multi_devices = [device for device in devices.values() if device.is_multi_device]

    if not multi_devices:
        console.print(
            Panel("[red]No multi-devices found.[/red]", title="Multi-Devices")
        )
        return

    # Create main tree
    tree = Tree(f"[bold blue]Multi-Devices ({len(multi_devices)} found)[/bold blue]")

    for device in multi_devices:
        # Create device branch
        device_text = (
            f"[green]{device.display_name}[/green] ([dim]{device.device_serial}[/dim])"
        )
        device_branch = tree.add(device_text)

        # Add device info
        device_branch.add(f"[yellow]Interface:[/yellow] {device.interface.value}")

        if device.channels:
            channels_branch = device_branch.add(
                f"[cyan]Channels ({len(device.channels)})[/cyan]"
            )

            for channel in device.channels.values():
                channel_type = channel.__class__.__name__
                location = ""
                if channel.room_name:
                    location = f" [dim]({channel.room_name})[/dim]"
                elif channel.floor_name:
                    location = f" [dim]({channel.floor_name})[/dim]"

                channel_text = f"{channel.channel_name}{location}"
                channel_branch = channels_branch.add(channel_text)
                channel_branch.add(f"[magenta]Type:[/magenta] {channel_type}")
                channel_branch.add(f"[magenta]ID:[/magenta] {channel.channel_id}")
        else:
            device_branch.add("[red]No channels found[/red]")

    console.print(tree)


def display_multi_devices_table(devices: dict):
    """Display multi-devices using rich table formatting."""
    console = Console()

    multi_devices = [device for device in devices.values() if device.is_multi_device]

    if not multi_devices:
        console.print(
            Panel("[red]No multi-devices found.[/red]", title="Multi-Devices")
        )
        return

    table = Table(
        title="Multi-Devices Overview", show_header=True, header_style="bold magenta"
    )
    table.add_column("Device", style="green", width=30)
    table.add_column("Serial", style="dim", width=20)
    table.add_column("Interface", style="yellow", width=15)
    table.add_column("Channels", style="cyan", justify="right")

    for device in multi_devices:
        table.add_row(
            device.display_name,
            device.device_serial,
            device.interface.value,
            str(len(device.channels)),
        )

    console.print(table)

    # Show detailed channel info
    for device in multi_devices:
        if device.channels:
            console.print(f"\n[bold green]{device.display_name}[/bold green] Channels:")

            channel_table = Table(show_header=True, header_style="bold cyan")
            channel_table.add_column("Name", style="white", width=25)
            channel_table.add_column("Type", style="magenta", width=20)
            channel_table.add_column("ID", style="dim", width=10)
            channel_table.add_column("Location", style="yellow", width=20)

            for channel in device.channels.values():
                location = channel.room_name or channel.floor_name or "Unknown"
                channel_table.add_row(
                    channel.channel_name,
                    channel.__class__.__name__,
                    channel.channel_id,
                    location,
                )

            console.print(channel_table)


if __name__ == "__main__":
    _fah = FreeAtHome(
        FreeAtHomeApi(
            host="http://<hostname or ip>/", username="installer", password="<password>"
        )
    )
    asyncio.run(_fah.load())

    _devices = _fah.get_devices()

    # Try different display methods
    print("=" * 80)
    print("TREE VIEW:")
    print("=" * 80)
    display_multi_devices_rich(_devices)

    print("\n" + "=" * 80)
    print("TABLE VIEW:")
    print("=" * 80)
    display_multi_devices_table(_devices)
```

kingsleyadam commented on custom_components/abbfreeathome_ci/binary_sensor.py:
```diff
@@ -163,12 +166,27 @@ async def async_will_remove_from_hass(self) -> None:
     @property
     def device_info(self) -> DeviceInfo:
         """Information about this entity/device."""
+        if self._create_subdevices and self._channel.device.is_multi_device:
+            return {
+                "identifiers": {
+                    (
+                        DOMAIN,
+                        f"{self._channel.device_serial}_{self._channel.channel_id}",
+                    )
+                },
+                "name": self._channel.channel_name,
+                "manufacturer": "ABB Busch-Jaeger",
+                "serial_number": f"{self._channel.device_serial}_{self._channel.channel_id}",
+                "suggested_area": self._channel.room_name,
+                "via_device": (DOMAIN, self._channel.device_serial),
+            }
+
         return {
             "identifiers": {(DOMAIN, self._channel.device_serial)},
             "name": self._channel.device_name,
-            "manufacturer": "ABB busch-jaeger",
+            "manufacturer": "ABB Busch-Jaeger",
             "serial_number": self._channel.device_serial,
-            "suggested_area": self._channel.room_name,
+            "suggested_area": self._channel.device.room_name,
             "via_device": (DOMAIN, self._sysap_serial_number),
         }
```
And this could look something like

```python
    @property
    def device_info(self) -> DeviceInfo:
        """Information about this entity/device."""
        if self._create_subdevices and self._channel.device.is_multi_device:
            return DeviceInfo(
                identifiers={
                    (
                        DOMAIN,
                        f"{self._channel.device_serial}_{self._channel.channel_id}",
                    )
                },
                name=self._channel.channel_name,
                manufacturer="ABB Busch-Jaeger",
                serial_number=f"{self._channel.device_serial}_{self._channel.channel_id}",
                suggested_area=self._channel.room_name,
                via_device=(DOMAIN, self._channel.device_serial),
            )

        return DeviceInfo(identifiers={(DOMAIN, self._channel.device_serial)})
```

derjoerg commented on custom_components/abbfreeathome_ci/__init__.py:
```diff
@@ -164,6 +166,20 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         configuration_url=entry.data[CONF_HOST],
     )
 
+    # Check, if sub-devices for multi-devices should be created
+    if entry.data[CONF_CREATE_SUBDEVICES]:
+        for _device in _free_at_home.get_devices().values():
+            if _device.is_multi_device:
+                device_registry.async_get_or_create(
+                    config_entry_id=entry.entry_id,
+                    identifiers={(DOMAIN, _device.device_serial)},
+                    name=_device.display_name,
+                    manufacturer="ABB Busch-Jaeger",
+                    serial_number=_device.device_serial,
+                    suggested_area=None,
+                    via_device=(DOMAIN, entry.data[CONF_SERIAL]),
+                )
```
With this code I also get Hue- and Sonos-devices created 😞

derjoerg replied:
Nope, I updated my local clone and always working with my local package 😉

derjoerg replied:
OK,
sorry for the confusion with the Hue- and Sonos-device. Forget this. It looks like I had some wrong caches.

But ...

This code also includes e.g. scenes:
```
      "FFFF48000002": {
        "floor": "02",
        "room": "07",
        "deviceId": "4800",
        "displayName": "RT AN",
        "unresponsive": false,
        "unresponsiveCounter": 0,
        "defect": false,
        "channels": {
          "ch0000": {
            "displayName": "RT AN",
            "floor": "02",
            "function": "FID_SCENE",
            "functionID": "4800",
            "inputs": {},
            "outputs": {
              "odp0000": {
                "pairing": "AL_SCENE_CONTROL",
                "pairingID": 4,
                "value": "1"
              }
            },
            "parameterNames": {},
            "parameters": {},
            "room": "07",
            "selectedIcon": "136"
          }
        },
        "parameters": {},
        "parameterNames": {}
      },
```

derjoerg replied:
Or this one:
```
      "ABB700000000": {
        "interface": "sysap",
        "deviceId": "FFFF",
        "displayName": "SysAP",
        "unresponsive": false,
        "unresponsiveCounter": 0,
        "defect": false,
        "channels": {
          "ch0000": {
            "displayName": "SysAP",
            "function": "FID_DAY_NIGHT_SENSOR",
            "functionID": "25",
            "inputs": {
              "idp0000": {
                "pairing": "AL_NIGHT_ACTUATOR_FOR_SYSAP",
                "pairingID": 20,
                "value": ""
              }
            },
            "outputs": {
              "odp0000": {
                "pairing": "AL_NIGHT",
                "pairingID": 18,
                "value": "1"
              }
            },
            "parameterNames": {},
            "parameters": {},
            "selectedIcon": "4"
          }
        },
        "parameters": {},
        "parameterNames": {}
      },
```

derjoerg commented:
I added a check for special F@H-devices. They are valid in the package but not useful (at least in the moment) for HA, so these devices will be skipped.

I just have one problem I need to look at, but it is too late, so I changed the PR back to draft.

kingsleyadam commented on custom_components/abbfreeathome_ci/__init__.py:
```diff
@@ -166,19 +166,24 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-                )
+    for _device in _free_at_home.get_devices().values():
+        # Skip special F@H devices
+        if (
+            _device.device_serial.startswith("FFFF")  # Scenes
+            or _device.device_id == "FFFF"  # System Access Point
+            or len(_device.channels.keys()) == 0
+        ):
+            continue
```
I like this, I'm wondering, would looking for the number of channels be enough? If they have a channel they'll likely have an entity.

In Python you can also use `_device.channels` to check if the dict is empty.

```suggestion
        if not _device.channels:
            continue
```

derjoerg replied:
No, looking for the number of channels isn't enough. E.g. Scenes always have one channel and the SysAP with a deviceId="FFFF" (serial="ABB700000000") also.

The problem I just faced with (but I think this is a local problem) is, that my configuration shows two "SmartTouch 10" panels. One I'm using (assigned to a room, door-bell event and so on) and another one (with a different serial) also with a lot of channels but not assigned to rooms. [And I don't have and never had two of these panels] This second panel shows up in the new code as an empty device. As stated: I think this is a local problem with my F@H configuration and has nothing to do with the code, but I just want to make some additional tests before I think the code is clean.

kingsleyadam replied:
I see one other scenario that could come up where a device could be added but all of it's channels are orphans. I'll add one additional function to the FreeAtHome class called `get_channels_by_device` which will pull in the filtered channels vs the channels directly from the device.

domi8671 replied:
Regarding the ghost device, your not alone. I have 2 Smarttouch 10 and after the last update 2 new appeared with different serials, I deleted them aswell.

kingsleyadam replied:
> No, looking for the number of channels isn't enough. E.g. Scenes always have one channel and the SysAP with a deviceId="FFFF" (serial="ABB700000000") also.
> 
> The problem I just faced with (but I think this is a local problem) is, that my configuration shows two "SmartTouch 10" panels. One I'm using (assigned to a room, door-bell event and so on) and another one (with a different serial) also with a lot of channels but not assigned to rooms. [And I don't have and never had two of these panels] This second panel shows up in the new code as an empty device. As stated: I think this is a local problem with my F@H configuration and has nothing to do with the code, but I just want to make some additional tests before I think the code is clean.

Can you give it another go with the new function I just added `get_channels_by_device`? If it returns any channels for a device then that channel will be added to an entity and will need a device. So I'm wondering what scenario exists that this isn't the case.

I adjusted my local code to use.

```pyhon
        if not _free_at_home.get_channels_by_device(_device.device_serial):
            continue
```

And now all of my devices also have at least 1 entity (e.g. no extra SysAP device)

kingsleyadam replied:
Also to note, I've compared my "live" Home Assistant instance with this branch and my dev instances and the number match-up exactly.

Granted, I don't have any many types of devices. But this is what I'd expect. If you have discrepancies after this change I'd like to know where those are at.
